### PR TITLE
Simulate tile server throughput with k6 load test

### PR DESCRIPTION
**Pull request by rbreslow**
## Overview

I looked through the CloudFront request logs for the last few outages using variations of the following Athena query:

```sql
WITH logs AS 
    (SELECT *,
         parse_datetime( concat( concat( format_datetime(date,
         'yyyy-MM-dd'), '-' ), time ),'yyyy-MM-dd-HH:mm:ss') AS datetime_ts
    FROM prd_cloudfront_logs
    WHERE date = date '2020-09-10')
SELECT time_taken, status, to_iso8601(datetime_ts) as datetime, request_ip, request_protocol || '://' || host_header || uri || '?' || url_decode(query_string) as url
FROM logs
WHERE datetime_ts
    BETWEEN timestamp '2020-09-10 12:24:00'
        AND timestamp '2020-09-10 12:35:00'
        ORDER BY datetime_ts ASC
```

Here is an excerpt of requests around the time that health checks timed out:

image

I observed that tile requests were being served in less than **half a second**, but it took **8 seconds** to serve the first tile with a query string attached. It took **24 tile requests** for overall request latency to climb to **30 seconds**, after which health checks timed out.

After being unsuccessful in trying to replay the exact requests exported from Athena, I attempted to simulate an incident by filtering by contributor `139` and zooming in on a random spot, Rio de Janerio. As soon as I zoomed in, the application stopped responding, and within **30 seconds** ECS terminated the tasks due to health check timeouts.

Connects #1084 

## Testing Instructions

Here is an example of the load test being ran with the same parameters as production (2 ECS tasks w/ 1 vCPU and 3 workers, `db.t3.small`).

```bash
running (01m39.1s), 00/10 VUs, 10 complete and 0 interrupted iterations
default ✓ [======================================] 10 VUs  01m38.9s/10m0s  10/10 shared iters


    ✗ is status 200
     ↳  15% — ✓ 15 / ✗ 80

    checks.....................: 15.78% ✓ 15   ✗ 80
    data_received..............: 331 kB 3.3 kB/s
    data_sent..................: 33 kB  336 B/s
  ✗ failed requests............: 84.21% ✓ 80   ✗ 15
    http_req_blocked...........: avg=3.86ms   min=2.72µs   med=5.44µs   max=57.99ms  p(90)=27.22ms  p(95)=27.95ms
    http_req_connecting........: avg=1.31ms   min=0s       med=0s       max=12.56ms  p(90)=10.62ms  p(95)=11.42ms
    http_req_duration..........: avg=43.95s   min=114.49ms med=59.97s   max=1m0s     p(90)=1m0s     p(95)=1m0s
    http_req_receiving.........: avg=69.82µs  min=0s       med=0s       max=592.2µs  p(90)=166.43µs p(95)=326.99µs
    http_req_sending...........: avg=122.86µs min=43.58µs  med=109.51µs max=390.14µs p(90)=169.59µs p(95)=230.21µs
    http_req_tls_handshaking...: avg=1.91ms   min=0s       med=0s       max=20.26ms  p(90)=14.85ms  p(95)=16.02ms
    http_req_waiting...........: avg=43.95s   min=114.26ms med=59.97s   max=1m0s     p(90)=1m0s     p(95)=1m0s
    http_reqs..................: 95     0.958233/s
    iteration_duration.........: avg=1m27s    min=1m5s     med=1m34s    max=1m38s    p(90)=1m38s    p(95)=1m38s
    iterations.................: 10     0.100867/s
    vus........................: 5      min=5  max=10
    vus_max....................: 10     min=10 max=10
```

The k6 cloud interface shows that request times steadily increase, as in the Athena logs, until they timeout.

image

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

### CODE REVIEW

**Comment by hectcastro** [comment]
Below is a table of results for a few different configuration combinations. Each of the tests attempted to execute 95 consecutive tile service requests via one (1) virtual user with distinct cache keys to help negate the throughput benefits of a CloudFront cache.

Changes to the test configuration applied can be seen in https://github.com/open-apparel-registry/open-apparel-registry/compare/804301b...96ae924.

| Option | Instances | Service vCPU | Service RAM (GB) | Database instance type | Gunicorn workers | Failed requests | Requests/sec | p95    |
| ------ | --------- | ---- | -------- | ---------------------- | ---------------- | --------------- | ------------ | ------ |
| 1      | 2         | 4    | 8        | `db.t3.small`          | 9                | 94              | 0.65         | 13.53s |
| 2      | 4         | 4    | 8        | `db.t3.small`          | 9                | 1               | 0.53         | 20.79s |
| 3      | 2         | 4    | 8        | `db.t3.medium`         | 9                | 0               | 0.72         | 12.73s |
| 4      | 4         | 4    | 8        | `db.t3.medium`         | 9                | 3               | 0.69         | 13.28s |
| 5      | 2         | 4    | 8        | `db.t3.large`          | 9                | 0               | 0.76         | 11.03s |
| 6      | 4         | 4    | 8        | `db.t3.large`          | 9                | 0               | 0.67         | 14.27s |
| 7      | 2         | 4    | 8        | `db.t3.xlarge`         | 9                | 0               | 1.52         | 5.86s  |
| 8      | 4         | 4    | 8        | `db.t3.xlarge`         | 9                | 0               | 1.38         | 6.77s  

Some general observations:

- It appears as though the most constant bottleneck is the database (and the tile service database queries it is attempting to satisfy)
- As the expensive queries begin to pile up, database query queuing time increases, which increases the overall latency of web requests
- However, not having enough service instances can lead to a high rate of failures, quickly (see **Option 1**)
- Things seem to get progressively better as the database instance type is increased, even if database vCPU count remains the same (see **Options 3-8**)
- Upgrading to a `db.t3.large` appears to provide the best value in terms of vertically scaling the database; the `db.t3.xlarge` is approximately twice the cost, but also reduces the p95 by half (see **Options 5-8**)

**Comment by rbreslow** [comment]
Before the modifications introduced in 804301b...96ae924, I was running a set of load tests. Below is a table of results. 

Each of the tests attempted to execute 95 consecutive tile requests, batching requests that occurred at the same time in the HAR file, and sleeping between each batch to replicate user interaction (e.g. zoom, wait a second, zoom again). I [invalidated](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Invalidation.html) the CloudFront cache between each test to remove that from the equation.

| Option | Instances | Service vCPU | Service RAM (GB) | Database instance type | Gunicorn workers | Failed requests | Requests/sec | Request duration (p95) |
|--------|-----------|--------------|------------------|------------------------|------------------|-----------------|--------------|------------------------|
| 1      | 2         | 1            | 8                | `db.t3.small`          | 3                | 80              | 0.95         | 1m0s                   |
| 2      | 8         | 1            | 8                | `db.t3.small`          | 3                | 88              | 1.45         | 59.98s                 |
| 3      | 2         | 4            | 8                | `db.t3.small`          | 9                | 95              | 0.87         | 59.97s                 |
| 4      | 2         | 1            | 8                | `db.t3.xlarge`         | 3                | 15              | 0.69         | 44.36s                 |
| 5      | 8         | 1            | 8                | `db.t3.xlarge`         | 3                | 3               | 1.25         | 53.33s                 |
| 6      | 2         | 4            | 8                | `db.t3.xlarge`         | 9                | 26              | 1.17         | 1m0s                   |

Some general observations:

- The strategy of scaling up the number of Gunicorn workers in an ECS task yielded the worst results on each database class (See **Options 3 and 6**).
- As Hector also observed, the database appears to be the most constant bottleneck.
- On the smallest database, increasing the ECS task count increases the amount of failures (See **Option 2**). However, increasing the ECS task count in addition to upgrading to a `db.t3.xlarge` actually decreases the amount of failures (See **Option 5**).
- I didn't consider testing on instances with the same vCPU count since CloudWatch showed the bottleneck to be entirely CPU Utilization. However, since one of our partners is testing staging today, I can't run additional tests just yet.
- It appears that upgrading the database to an instance with more vCPUs is the best way to work around the problem. Doubling the vCPUs with the jump from a `db.t3.small` to a `db.t3.xlarge` led to an 82% decrease in failed requests (96% with **Option 5**). However, going off of on-demand RDS pricing, our database cost would increase 8x. If we tweaked the ECS task count, Fargate cost would increase 4x. It's possible AWS Savings Plan + RI coverage could help mitigate some of the cost.

**Comment by hectcastro** [comment]
I think using the simplified version of the test is valuable because it allows the services to run with less failures, which provides a greater chance to extract signal vs. noise. We can always reevaluate those outcomes against the more aggressive version of the test afterwards.

It seems pretty clear that the database is the bottleneck, and that settling on either a `db.t3.large` or `db.t3.xlarge` is a recommendation that puts us in a better position to satisfy requests. What's less clear to me is the sweet spot of web request responding units (either more service instances with lower Gunicorn worker counts, or less service instances with higher Gunicorn worker counts). For example, I kept mine constant at 9, but if I had more time, I would have liked to try lower counts. You have more of a mix, but the high failure rate makes it harder to feel confident in coming to strong conclusions.

I think there could be value in doing another round of tests that keeps the cardinality of certain variables low, but tries to identify that sweet spot. For example, only target `db.t3.large` and `db.t3.xlarge` database instance types, and only allocate 2 vCPUs per service instance, but test 2-4 combinations of service instances along with 2-4 combinations of Gunicorn worker counts.

Another option could be to just stick with the database instance type increases as the primary recommendation and leave the other variables as-is. We can let that run as an experiment for a sprint to see how the number of outages compares, but also commit to reviewing resource utilization and access logs to see where access patterns came close to causing problems, but didn't.

**Comment by rbreslow** [comment]
I ran another set of load tests to see how making the tweaks outlined in #1091 would effect performance.

**Options 1-3** use a 1:1 ratio of Gunicorn workers to vCPUs.

| Option | Instances | Service vCPU | Service RAM (GB) | Database instance type | Gunicorn workers | Failed requests | Requests/sec | Request duration (p95) |
|--------|-----------|--------------|------------------|------------------------|------------------|-----------------|--------------|------------------------|
| 1      | 2         | 1            | 8                | `db.t3.xlarge`         | 1                | 87              | 1.25         | 59.98s                 |
| 2      | 4         | 1            | 8                | `db.t3.xlarge`         | 1                | 79              | 0.81         | 59.98s                 |
| 3      | 8         | 1            | 8                | `db.t3.xlarge`         | 1                | 71              | 1.23         | 59.98s                 |


**Options 4-6** use a 4:4 ratio of Gunicorn workers to vCPUs.

| Option | Instances | Service vCPU | Service RAM (GB) | Database instance type | Gunicorn workers | Failed requests | Requests/sec | Request duration (p95) |
|--------|-----------|--------------|------------------|------------------------|------------------|-----------------|--------------|------------------------|
| 4      | 2         | 4            | 8                | `db.t3.xlarge`         | 4                | 47              | 0.93         | 59.98s                 |
| 5      | 4         | 4            | 8                | `db.t3.xlarge`         | 4                | 25              | 0.69         | 1m0s                   |
| 6      | 8         | 4            | 8                | `db.t3.xlarge`         | 4                | 16              | 0.83         | 59.98s                 |

These results actually look fairly _worse_ than the initial load tests I ran, where we left the Gunicorn worker formula alone (`(2 x vCPUs) + 1 `) and simply horizontally scaled the task instances.

At a high level, vertically scaling the database seems to provide the biggest boost, while horizontally scaling the number of ECS tasks (See **Option 5** in the original load test) appears to be the best tweak we can make on the application side.

**Comment by hectcastro** [comment]
Did you get a chance to test PostgreSQL 12?

**Comment by hectcastro** [comment]
Please try taking another pass with your original load test configuration against staging with https://github.com/open-apparel-registry/open-apparel-registry/pull/1109 applied to see if you can reproduce results similar to these below:

```
    checks.....................: 100.00% ✓ 95   ✗ 0
    data_received..............: 382 kB  24 kB/s
    data_sent..................: 32 kB   2.0 kB/s
  ✓ failed requests............: 0.00%   ✓ 0    ✗ 95
    http_req_blocked...........: avg=3.2ms    min=100ns    med=300ns    max=76.77ms  p(90)=13.84ms  p(95)=25.51ms
    http_req_connecting........: avg=1.2ms    min=0s       med=0s       max=13.81ms  p(90)=4.96ms   p(95)=11.5ms
    http_req_duration..........: avg=390.17ms min=173.87ms med=342.48ms max=857.44ms p(90)=593.98ms p(95)=649.76ms
    http_req_receiving.........: avg=77.33µs  min=18.1µs   med=44.8µs   max=992.1µs  p(90)=97.62µs  p(95)=261.18µs
    http_req_sending...........: avg=52.99µs  min=15.8µs   med=42.1µs   max=325.7µs  p(90)=73.86µs  p(95)=101.44µs
    http_req_tls_handshaking...: avg=1.68ms   min=0s       med=0s       max=39.11ms  p(90)=7.21ms   p(95)=13.39ms
    http_req_waiting...........: avg=390.04ms min=173.75ms med=342.39ms max=857.32ms p(90)=593.91ms p(95)=649.68ms
    http_reqs..................: 95      6.037107/s
    iteration_duration.........: avg=6.49s    min=1.3s     med=5.03s    max=15.67s   p(90)=11.08s   p(95)=13.38s
    iterations.................: 10      0.635485/s
    vus........................: 1       min=1  max=10
    vus_max....................: 10      min=10 max=10
```

File: deployment/terraform/dns.tf [hunk 1]
```diff
@@ -20,7 +20,7 @@ resource "aws_route53_record" "database" {
   name    = "database.service.${var.r53_private_hosted_zone}"
   type    = "CNAME"
   ttl     = "10"
-  records = ["${module.database_enc.hostname}"]
+  records = ["openapparelregistry-prd-clone.ctdp4povekhv.eu-west-1.rds.amazonaws.com"]
```
Comment (hectcastro) [inline deployment/terraform/dns.tf:23]: Reminder to revert prior to merge.

**Review by hectcastro — APPROVED** [review]
👍

**Comment by rbreslow** [comment]
> Please try taking another pass with your original load test configuration against staging with #1109 applied to see if you can reproduce results similar to these below:

Wow!!! Awesome!!! My results looked identical. I loaded up the application in my browser and tried to drive it hard. It's not _super_ responsive, but **it works**. 

Also, I scaled the database back down to a `db.t3.small`. It works in my browser, and here are the results of a load test:

```bash
    checks.....................: 100.00% ✓ 95   ✗ 0
    data_received..............: 381 kB  24 kB/s
    data_sent..................: 32 kB   2.0 kB/s
  ✓ failed requests............: 0.00%   ✓ 0    ✗ 95
    http_req_blocked...........: avg=11.11ms  min=2.86µs   med=5.44µs   max=193.69ms p(90)=72.21ms  p(95)=75.01ms
    http_req_connecting........: avg=3.94ms   min=0s       med=0s       max=35.01ms  p(90)=33.36ms  p(95)=34.17ms
    http_req_duration..........: avg=1.07s    min=229.77ms med=1.04s    max=1.95s    p(90)=1.58s    p(95)=1.71s
    http_req_receiving.........: avg=143.31µs min=44.21µs  med=102.18µs max=1.79ms   p(90)=146.68µs p(95)=371.77µs
    http_req_sending...........: avg=119.64µs min=40.92µs  med=121.45µs max=439.1µs  p(90)=166.57µs p(95)=180.79µs
    http_req_tls_handshaking...: avg=6.49ms   min=0s       med=0s       max=129.38ms p(90)=37.78ms  p(95)=40.37ms
    http_req_waiting...........: avg=1.07s    min=229.48ms med=1.04s    max=1.95s    p(90)=1.58s    p(95)=1.71s
    http_reqs..................: 95      5.907857/s
    iteration_duration.........: avg=7.96s    min=2.98s    med=6.63s    max=15.88s   p(90)=11.86s   p(95)=13.87s
    iterations.................: 10      0.62188/s
    vus........................: 0       min=0  max=10
    vus_max....................: 10      min=10 max=10
```

**Comment by rbreslow** [comment]
I upgraded the database to PostgreSQL 12, still on a `db.t3.small`, just to see how that would influence the results. 

I lost the logs for the first load test, but the iteration duration was around **~30 seconds**. I believe the database was just warming up because subsequent load tests were much better.

<details>   
<summary>1</summary>    

```bash
    checks.....................: 100.00% ✓ 95   ✗ 0
    data_received..............: 385 kB  24 kB/s
    data_sent..................: 32 kB   1.9 kB/s
  ✓ failed requests............: 0.00%   ✓ 0    ✗ 95
    http_req_blocked...........: avg=10.39ms  min=2.93µs   med=5.37µs   max=158.78ms p(90)=73.33ms  p(95)=76.29ms
    http_req_connecting........: avg=4.01ms   min=0s       med=0s       max=37.01ms  p(90)=33.64ms  p(95)=34.4ms
    http_req_duration..........: avg=1.3s     min=220.45ms med=1.44s    max=2.36s    p(90)=2.02s    p(95)=2.12s
    http_req_receiving.........: avg=179.6µs  min=57.76µs  med=95.68µs  max=2.32ms   p(90)=290.46µs p(95)=533.19µs
    http_req_sending...........: avg=117.29µs min=40.43µs  med=113.35µs max=292.43µs p(90)=159.77µs p(95)=199.63µs
    http_req_tls_handshaking...: avg=5.35ms   min=0s       med=0s       max=106.75ms p(90)=36.84ms  p(95)=41.46ms
    http_req_waiting...........: avg=1.3s     min=220.3ms  med=1.44s    max=2.36s    p(90)=2.02s    p(95)=2.12s
    http_reqs..................: 95      5.79664/s
    iteration_duration.........: avg=8.44s    min=3.78s    med=7.38s    max=16.2s    p(90)=11.9s    p(95)=14.05s
    iterations.................: 10      0.610173/s
    vus........................: 1       min=1  max=10
    vus_max....................: 10      min=10 max=10
```

</details>

<details>   
<summary>2</summary>    

```bash
    checks.....................: 100.00% ✓ 95   ✗ 0
    data_received..............: 381 kB  24 kB/s
    data_sent..................: 32 kB   2.0 kB/s
  ✓ failed requests............: 0.00%   ✓ 0    ✗ 95
    http_req_blocked...........: avg=3.21ms   min=2.58µs   med=5.09µs   max=98.3ms   p(90)=17.86ms  p(95)=19.08ms
    http_req_connecting........: avg=700.12µs min=0s       med=0s       max=6.84ms   p(90)=4.75ms   p(95)=5.45ms
    http_req_duration..........: avg=1.18s    min=216.27ms med=1.19s    max=2.35s    p(90)=1.88s    p(95)=2.04s
    http_req_receiving.........: avg=125.21µs min=40.5µs   med=97.43µs  max=958.53µs p(90)=147.3µs  p(95)=346.26µs
    http_req_sending...........: avg=114.19µs min=44.83µs  med=110.91µs max=598.49µs p(90)=143.13µs p(95)=189.47µs
    http_req_tls_handshaking...: avg=2.31ms   min=0s       med=0s       max=76.89ms  p(90)=12.34ms  p(95)=13.3ms
    http_req_waiting...........: avg=1.18s    min=216.07ms med=1.19s    max=2.35s    p(90)=1.88s    p(95)=2.04s
    http_reqs..................: 95      6.01092/s
    iteration_duration.........: avg=8.13s    min=2.98s    med=7.1s     max=15.62s   p(90)=11.88s   p(95)=13.75s
    iterations.................: 10      0.632728/s
    vus........................: 1       min=1  max=10
    vus_max....................: 10      min=10 max=10
```

</details>

p95s across the board look to be a bit lower and the application feels more responsive in my browser.

**Pull request merged**